File: tests/upstream/adapters/mcp_call.py
#!/usr/bin/env python3
"""Thin Python-boundary adapter: one MCP tools/call over stdio, no assertions.

Usage:
  FM_HOME=/tmp/stub python3 tests/upstream/adapters/mcp_call.py fleet_snapshot '{}'
"""
import json
import os
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent.parent.parent
SERVER = ROOT / "fm_mcp_server.py"


def main():
    if len(sys.argv) != 3:
        print("usage: mcp_call.py <tool> '<json-args>'", file=sys.stderr)
        return 2
    tool, args = sys.argv[1], json.loads(sys.argv[2])
    proc = subprocess.Popen(
        ["python3", str(SERVER)],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
        text=True,
        bufsize=1,
        env=dict(os.environ),
    )
    assert proc.stdin is not None and proc.stdout is not None
    proc.stdin.write(json.dumps({"jsonrpc": "2.0", "id": 1, "method": "tools/call",
                                 "params": {"name": tool, "arguments": args}}) + "\n")
    proc.stdin.flush()
    line = proc.stdout.readline()
    print(line.strip())
    try:
        proc.stdin.close()
    except Exception:
        pass
    try:
        proc.wait(timeout=10)
    except Exception:
        proc.kill()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/upstream/adapters/mcp_call.mjs
#!/usr/bin/env node
/**
 * Thin TypeScript-boundary adapter: one MCP tools/call over stdio, no assertions.
 * The server spawns via process.execPath, so `node mcp_call.mjs` proves node
 * and `bun mcp_call.mjs` proves bun.
 *
 * Usage:
 *   FM_HOME=/tmp/stub node tests/upstream/adapters/mcp_call.mjs fleet_snapshot '{}'
 */
import { spawn } from "node:child_process";
import path from "node:path";
import readline from "node:readline";
import { fileURLToPath } from "node:url";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(HERE, "..", "..", "..");
const SERVER = path.join(ROOT, "ts", "dist", "server.js");

const [tool, argsJson] = process.argv.slice(2);
if (!tool || argsJson === undefined) {
  console.error("usage: mcp_call.mjs <tool> '<json-args>'");
  process.exit(2);
}

const child = spawn(process.execPath, [SERVER], {
  stdio: ["pipe", "pipe", "ignore"],
  env: { ...process.env },
});
const rl = readline.createInterface({ input: child.stdout });
child.stdin.write(
  JSON.stringify({ jsonrpc: "2.0", id: 1, method: "tools/call", params: { name: tool, arguments: JSON.parse(argsJson) } }) + "\n",
);
const line = await new Promise((resolve) => rl.once("line", resolve));
console.log(line);
child.stdin.end();
await new Promise((resolve) => {
  const timer = setTimeout(() => { try { child.kill("SIGKILL"); } catch {} resolve(); }, 10000);
  child.on("exit", () => { clearTimeout(timer); resolve(); });
});
rl.close();
